l3backend/l3backend-color.dtx

```diff
@@ -1552,22 +1552,32 @@ end
 %    \end{macrocode}
 % \end{macro}
 %
-% \begin{macro}{\tex_pdfxform:D}
-%   We redefine the \pkg{expl3} version of the primitive to do the replacement
-%   implicitly. The keywords \texttt{margin} and \texttt{type} are not
-%   documented at the \TeX{} level.
+%<@@=pdfxform>
+%
+% \begin{macro}{pdfxform}
+% \begin{macro}{\@@_backend_now:w, \@@_backend_shipout:w}
```

Comment:
I think the names are wrong in this context. The difference between immediate and non-immediate xform are not related to shipout. An non-immediate xform only reserve an object number and are kept in memory and only moved to the pdf when referenced with `\pdfrefxform`

Reply:
Hmm, to me that suggests that someone (Hans?) made a bit of a mis-step in having it take any notice of `\immediate` in the first place: that's _supposed_ to be about whether to wait for shipout or not.

Reply:
well it is the same with pdfobjects. There are only written when referenced or if you use \immediate. It is a bit confusing that both use the \immediate command to distinguish the case, which is normally connected to "directly or at shipout", but probably they used it as it was there, and why invent a new command name?

Reply:
@davidcarlisle Suggested that we look in the 2e kernel at a full set of macro interfaces for handling `\immediate` vs 'later' in a way that's easier to integrate with LuaTeX. On that basis, I think I would give macros for 'now or when used' different names from 'now or shipout'. Something like 'firstuse'?

Reply:
`now` and `delayed`? We could then keep `\pdf_object_write:nn` as the "now" variant and add `\pdf_object_write_delayed:nn` and similar for xform. Imho it would be useful, it happens that one doesn't know if an object will be really needed at the end, for example for namespaces.

Reply:
> add `\pdf_object_write_delayed:nn` and similar for xform.

I think the names for objects and xforms have to reflect that they behave quite differently since objects are "only" written out, while the xform equivalent will also always reference the xform in the the current page stream IIRC.

Reply:
At the moment this is all strictly internal stuff - although I agree we should try to get to a position that can go public

Reply:
@zauguin objects and xform are different, but I think that is reflected in the start of the name `\pdf_object` and `\pdfxform`,  the behaviour "delayed" part is imho similar enough to justify the same word. In both cases the "object" hangs around in the limbo and is dropped if not used, and when used both behave as if the immediate version has been used there.